flexible upload support

rnicholus opened the pull request:
The following issues will be addressed once this is released:

- #848 
- #1697 
- #1755 
- #1325 
- #1647 
- #1703

Latest release is 5.16.0-alpha.16 (on npm).

Commit:
ports 3000/3001 clash with my local env, and possibly others
- moving to 4000/4001

Commit:
Start of support for promissory onUploadChunk event
- Manual tests w/ chunking and concurrent chunking look good.

TODO:
- handle resolve result
- handle failure
- unit tests
- docs uppdate
- typescript defs update

Commit:
returned onUploadChunk promise con override method, params, headers, & url

Commit:
always ensure test server are killed either on test start or stop

Commit:
don't try to kill test server on CI before tests start

Commit:
forgot to start the test resources server on CI

Commit:
promissory onUpload callback + handle async failure of onUpload/chunk

Commit:
option to allow upload responses without { "success": true }

Commit:
allow default params to be omitted from upload requests

Commit:
1st alpha release with most of the changes obviously needed for flexible workflow

Commit:
don't fail upload w/ non-JSON response when requireSuccessJson = false

Commit:
5.16.0-alpha.1

Commit:
more flexible chunking.success request support

Commit:
5.16.0-alpha.2

rnicholus commented:
Progress here:

- [x] `onUploadChunk` can return a `Promise` that optionally overrides endpoint, params, headers, and method for the chunk upload.
- [x] `onUpload` can return a `Promise`.
- [x] New function type options for the `chunking.success` option: `headers` and `params`. Also `jsonPayload` boolean and `method` string. Finally, `endpoint` accepts a function value. All function options are passed the file ID.
- [x] New `request.omitDefaultParams` option. If set, Fine Uploader doesn't send any "qq" params w/ the upload request.
- [x]  New `request.requireSuccessJson` option. If not set, upload endpoint does not need to return `{ "success": true }` to indicate success. 
- [ ] More flexible resume support (allow custom keys/data) 
- [ ] Unit tests to cover everything.
- [ ] Documentation updates.
- [ ] Document `chunking.mandatory` option.
- [ ] Typescript def file updates. (cc: @SinghSukhdeep)
- [ ] fine-uploader-wrappers updates to cover any new API features (i.e. more async support).

webmaster-mms commented:
Hi Ray Nicholus, my most sincere compliments for the work done. I'm starting to use fine-uploader, I installed everything properly and is working. I do not know how to pass photos to a db on the site do not have any tools or demos? thank you very much.

rnicholus committed:
Merge branch 'develop' into feature/flexible-upload-support

Commit:
make sure test servers are never cleaned up before tests run

Commit:
add .editorconfig (can't believe this didn't exist until now)

Commit:
Allow custom resume keys and data to be specified.
fixes #1755

Commit:
include customResumeData in return value of getResumableFilesData API method

Commit:
add isResumable public API method

Commit:
obey the lint

Commit:
onUpload rejected Promise does not fail upload

Commit:
onUpload rejected Promise does not fail upload after all auto-retries are exhausted

Commit:
introduce chunking.success.resetOnStatus to allow FU to reset a file based on chunking.success response code

Commit:
new API method: isResumable(id)

Commit:
Allow onUpload resolved Promise to pause the file.
Use case: When onUpload is called, you make a request to your server to see if the file already exists. If it does, you want to let your user decide if they want to overwrite the file, or cancel the upload entirely. While waiting for user input you don't want to hold a spot in the upload queue. If the user decided to overwrite the file, call the `continueUpload` API method.
closes #848

Commit:
Allow per-file chunk sizes to be specified.
chunking.partSize now accepts a function, which passes the file ID and size
closes #1697

rnicholus commented:
All planned development work for this PR is now complete. The following issues will be addressed once this is released:

- #848 
- #1697 
- #1755 
- #1325 
- #1647 

The current release with these changes (pushed to npm) is v5.16.0-alpha.12. The next step is to regroup, write some unit tests, update the docs, and upgrade this to a beta release if all goes well. After I've been using all of these changes in production for a bit I'll release 5.16.0.

Commit:
unit tests for more flexible chunking.success support

Commit:
unit tests for new request.requestJsonSuccess and request.omitDefaultParams config options

Commit:
tests for custom resume data/keys & chunking.success.resetOnStatus

Commit:
onUpload thenable support tests

SinghSukhdeep commented:
@rnicholus I can work on updating the TS def now? as your planned development is complete.

rnicholus commented:
@SinghSukhdeep sounds good. I don't have any of the docs updated yet, so if anything isn't clear, LMK. there are a lot of changes here...

rnicholus committed:
Merge branch 'develop' into feature/flexible-upload-support

Commit:
onUploadChunk thenable support tests

Commit:
test for per-file chunk size support

rnicholus commented:
I believe I have now written sufficient unit tests for all changes/additions in this PR.

Commit:
release 5.16.0-alpha.13

Commit:
feat(beforeUnload): new option to turn off beforeUnload alert during uploads

Commit:
feat(features.js): auto-detect folder support
closes #1703

Commit:
Allow access to Blob when file status is still SUBMITTING

SinghSukhdeep commented:
Would need the docs updated before I can start with TS update.
Lot seems to have changed and I can't seem to figure everything out by looking at all the changes.
Doc update would point me in straight direction.

rnicholus commented:
Docs are next on my list...

Commit:
docs: options, API, and events doc updates

rnicholus commented:
API docs updated!

SinghSukhdeep commented on docs/api/events.jmd:
```diff
@@ -439,7 +444,13 @@ Useful for implementing an aggregate progress bar.
-{{ api_event("upload", "onUpload", "Called just before an item begins uploading to the server.",
+{{ api_event("upload", "onUpload",
+"""
+Called just before an item begins uploading to the server. In asynchronous work must be performed,
```
@rnicholus We are still talking about returning Fine-uploader's own implementation of Promise and not native Promise?
Also in this line, should it be `In asynchronous` or `If asynchronous`?

rnicholus replied:
any "thenable" object is permitted

rnicholus replied:
though i dont want to encourage or promote Fine Uploader's promise impl at all. it's non-standard and just plain bad

SinghSukhdeep replied:
I'll keep the ability to return both ATM then.
Is the `In` supposed to be `If` there?

rnicholus replied:
I'd really prefer we don't even mention `qq.Promise` anymore. I expect/hope to rip it out entirely in 6.0.

Yes, it should be "if" instead of "in".

SinghSukhdeep replied:
Ok that's fine. I'll make changes accordingly then.

singhjusraj committed:
TypeScript Definitions Updated

SinghSukhdeep commented:
TypeScript Updated

rnicholus commented:
Thanks @SinghSukhdeep. Looks like everything is done here. Just need to let all of these changes cook on a project of mine before I release (along with a bit more manual testing). Would love to hear if others are having success with this branch.

rnicholus committed:
Merge branch 'develop' into feature/flexible-upload-support

rnicholus committed:
Merge branch 'develop' into feature/flexible-upload-support

rnicholus committed:
Merge branch 'develop' into feature/flexible-upload-support
